gameState.cpp

```diff
@@ -2,82 +2,320 @@
 #include <iostream>
 #include <random>
 
-gameState::gameState(int numPlayers)
-    : numPlayers(numPlayers),
-      players(numPlayers, player()), deck{card::baron,       card::baron,
-                                          card::baron,       card::executioner,
-                                          card::executioner, card::executioner,
-                                          card::diplomat,    card::diplomat,
-                                          card::diplomat,    card::commander,
-                                          card::commander,   card::commander,
-                                          card::bodyguard,   card::bodyguard,
-                                          card::bodyguard} {
-
-  for (size_t i = 0; i < deck.size(); i++) {
-    std::cout << deck[i] << "\n";
-  }
+gameState::gameState(std::vector<std::unique_ptr<Player>> playerList)
+  : players(std::move(playerList)),
+                                     deck{Card::baron,       Card::baron,
+                                          Card::baron,       Card::executioner,
+                                          Card::executioner, Card::executioner,
+                                          Card::diplomat,    Card::diplomat,
+                                          Card::diplomat,    Card::commander,
+                                          Card::commander,   Card::commander,
+                                          Card::bodyguard,   Card::bodyguard,
+                                          Card::bodyguard} {
 
   // Shuffle deck
   std::cout << "-- Shuffle Deck Begin -- \n";
   auto rng = std::default_random_engine(time(0));
   std::shuffle(deck.begin(), deck.end(), rng);
-  for (size_t i = 0; i < deck.size(); i++) {
-    std::cout << deck[i] << "\n";
-  }
 
   // Create X numbers of players and distribute cards
   // Push top of deck into each players hands
 
-  for (int i = 0; i < numPlayers; i++) {
+  for (size_t i = 0; i < playerList.size(); i++) {
     std::cout << "Player " << i << " added. \n";
-    card first, second;
+    Card first, second;
     first = deck.back();
     deck.pop_back();
     second = deck.back();
     deck.pop_back();
-    players.at(i).hand = hand(first, second);
-    std::cout << "Player " << i << " : <" << players.at(i).hand.first << ","
-              << players.at(i).hand.second << "> \n";
+    players.at(i)->setHand(Hand(first, second));
   }
 }
 
-void gameState::mainLoop() {
-  // Loop through the player array and get them to do actions
-  // Probability is just randomized.
- while (1) {
-    if (players.size() == 1) {
-      break;
-    }
+Player *gameState::getPlayer(std::string name) {
+  for (auto &player : players)
+    if (player->name == name)
+      return player.get();
+  return nullptr;
+}
+
+void gameState::generalIncome(std::string name) {
+  getPlayer(name)->setCoin(getPlayer(name)->getCoin() + 1);
```

Comment:
I'd suggest using a local variable: Player *p = getPlayer(name); then use that rather than having a bunch of calls to getPlayer(name).